### DISCUSSION TITLE
Defining porosity, tortuosity, and particle size as a function of x

### DISCUSSION THREAD

**Discussion by amirsh213** [pybamm-team/PyBaMM#2108] [Q&A]
Hi,
I saw in one of the issues that people tried to define porosity and particle size as a function of x but I have not seen any instruction on how to do that. I defined the porosity as a function x using [data] and a CSV file but when I plot it after simulation, the interpolation function shifts the porosity profile downward meaning I have a smaller average porosity. 
Also, I was wondering how I can define the MacMullin number or tortuosity with a customized function (function of porosity) instead of the Bruggeman equation?
Thanks.

**Comment by valentinsulzer** [comment]
Can you post the code you ran and the results you got for x-dependent porosity?

For the non-Bruggeman tortuosity, we need to add that as a separate submodel, please open an issue

**Comment by amirsh213** [comment]
It is just a simple 1C discharge. After changing my parameters to desired values I used the attached porosity profile for the anode via [data]. I plotted both porosity profiles on one plot and it seems that Pybamm interpolation is having a hard time interpolating at both ends and skips the last 2.5 um at both ends although I defined my anode length as 98 um. 

```
parameter_values = pybamm.ParameterValues(chemistry=chemistry)

parameter_values["Initial concentration in negative electrode [mol.m-3]"] = 28000
parameter_values["Initial concentration in positive electrode [mol.m-3]"] = 7000

model = pybamm.lithium_ion.DFN()

solver= pybamm.CasadiSolver(mode="safe")

sim = pybamm.Simulation(model, parameter_values=parameter_values, solver=solver)

sim.solve([0, 3110])
sim.plot(['Negative electrode porosity'])

```
[porosity_profile.csv](https://github.com/pybamm-team/PyBaMM/files/8928905/porosity_profile.csv)
![porosity profiles](https://user-images.githubusercontent.com/77367621/174342220-542516f9-043c-4dc7-a948-b2e9ee698c2e.jpg)

**Reply by valentinsulzer** [reply]
It just doesn't have a grid point there. With the `FiniteVolume` discretization, grid points are in the center of nodes, and the mesh is such that the left edge of the first node is at the left edge of the domain, and the right edge of the last node is at the right of the domain. So the center of the first and last nodes are offset from the edge of the domain.

The easiest thing to do would be to increase the number of grid points (see https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/Getting%20Started/Tutorial%209%20-%20Changing%20the%20mesh.ipynb), as you add grid points you should see the two ends being captured. You can then do a convergence study on the outputs of interest (plot the output with various grid sizes) to see how many grid points you need in order to get an accurate enough result.

**Reply by RuiheLi** [reply]
Dear PyBaMM team,

May I ask are there any simple lines to do this? I have the same requirement and encounter some problems when I was trying to do:

`
import pybamm as pb;
V_max = 4.2;        V_min = 2.5; 
exp_RPT_text = [ (
    f"Discharge at 1C until {V_min} V",  
    "Rest for 1 hours (5 minute period)",  
    f"Charge at 1C until {V_max} V" ) ] 
Experiment_RPT    = pb.Experiment( exp_RPT_text * 2     ) 
Model_0 = pb.lithium_ion.DFN() 
Chemistry=pb.parameter_sets.OKane2022
Para_0=pb.ParameterValues(chemistry=Chemistry)
x_n = pb.standard_spatial_vars.x_n 
Para_0.update({'Negative electrode porosity': 0.25*x_n})
print(Para_0['Negative electrode porosity'])
Sim_0    = pb.Simulation(
        Model_0,        experiment = Experiment_RPT,
        parameter_values = Para_0,
        solver = pb.CasadiSolver(),) 
Sol_0    = Sim_0.solve(calc_esoh=False)
`

The error message is:
`TypeError: Parameter provided for 'Negative electrode porosity' is of the wrong type (should either be scalar-like or callable)`

**Reply by valentinsulzer** [reply]
try
```python3
def neg_porosity(x):
   return 0.25 * x
Para_0.update({'Negative electrode porosity': neg_porosity})
```

**Reply by RuiheLi** [reply]
Works! Thank you very much!

**Reply by RuiheLi** [reply]
Sorry for the spamming, may I further ask how to defien a porosity distribution that is lower at the negative electrode-seperator interface but higher at the negative electrode-current collector interface? I tried something like 
```
def neg_porosity(x):
   return 0.25 * (1-x)
Para_0.update({'Negative electrode porosity': neg_porosity})
```
but that doesn't work.

**Reply by valentinsulzer** [reply]
x is in dimensional units (micro meters)

**Reply by RuiheLi** [reply]
Thanks for the patience! Paste the successful scirpt here for anyone who may need it:

```
import pybamm as pb;
exp_RPT_text = [ (  "Discharge at 1C for 10 minutes", ) ] 
Experiment_RPT = pb.Experiment( exp_RPT_text * 1 ) 
Model_0 = pb.lithium_ion.DFN() 
Chemistry=pb.parameter_sets.OKane2022 
Para_0=pb.ParameterValues(chemistry=Chemistry) 
def neg_porosity(x_n):
   return 0.25 * (1-x_n/Para_0["Negative electrode thickness [m]"]) 
Para_0.update({'Negative electrode porosity': neg_porosity})
Sim_0 = pb.Simulation( 
    Model_0, 
    experiment = Experiment_RPT, 
    parameter_values = Para_0, 
    solver = pb.CasadiSolver(),) 
Sol_0 = Sim_0.solve(calc_esoh=False)
```
